docs/examples/move/nft_marketplace/sources/nft_marketplace.move

```diff
@@ -0,0 +1,117 @@
+module nft_marketplace::nft_marketplace {
+    // iota imports
+    use iota::{
+        kiosk::{Kiosk, KioskOwnerCap, purchase},
+        kiosk_extension,
+        bag,
+        transfer_policy::{Self, TransferPolicy, TransferPolicyCap, has_rule},
+        coin::Coin,
+        iota::IOTA,
+    };
+
+    // rules imports
+    // use kiosk::floor_price_rule::Rule as FloorPriceRule;
+    // use kiosk::personal_kiosk_rule::Rule as PersonalRule;
+    use kiosk::royalty_rule::Rule as RoyaltyRule;
+    use kiosk::royalty_rule;
+    // use kiosk::witness_rule::Rule as WitnessRule;
+
+
+    // === Errors ===
+    const EExtensionNotInstalled: u64 = 0;
+    const EObjectNotExist: u64 = 1;
+
+    // === Constants ===
+    const PERMISSIONS: u128 = 11;
+
+    /// Extension Key for Kiosk Marketplace extension.
+    public struct Marketplace has drop {}
+
+    /// Used as a key for the item that has been up for sale that's placed in the Extension's Bag.
+    public struct Listed has store, copy, drop { id: ID }
+    
+    public struct ItemPrice<T: key + store> has store {
+        /// Total amount of time offered for renting in days.
+        price: u64,
+    }
+
+    /// Enables someone to install the Marketplace extension in their Kiosk.
+    public fun install(
+        kiosk: &mut Kiosk,
+        cap: &KioskOwnerCap,
+        ctx: &mut TxContext,
+    ) {
+        kiosk_extension::add(Marketplace {}, kiosk, cap, PERMISSIONS, ctx);
+    }
+
+    /// Remove the extension from the Kiosk. Can only be performed by the owner,
+    /// The extension storage must be empty for the transaction to succeed.
+    public fun remove(kiosk: &mut Kiosk, cap: &KioskOwnerCap, _ctx: &mut TxContext) {
+        kiosk_extension::remove<Marketplace>(kiosk, cap);
+    }
+
+    public fun setup_royalties<T: key + store>(policy: &mut TransferPolicy<T>, cap: &TransferPolicyCap<T>, amount_bp: u16, min_amount: u64, ctx: &mut TxContext) {
+        royalty_rule::add<T>(policy, cap, amount_bp, min_amount);
+    }
+
+    /// Buy listed item and pay royalties if needed
+    public fun buy_item<T: key + store>(kiosk: &mut Kiosk, policy: &mut TransferPolicy<T>, item_id: object::ID, mut payment: Coin<IOTA>, ctx: &mut TxContext) {
+        assert!(kiosk_extension::is_installed<Marketplace>(kiosk), EExtensionNotInstalled);
+        let item_price = take_from_bag<T, Listed>(kiosk,  Listed { id: item_id });
+        let ItemPrice { price } = item_price;
+        let payment_amount = payment.split(price, ctx);
```

Comment:
For composability reasons, it is better to split the coin previously in the PTB and here check that `payment_amount_value == price + royalties_value`

Reply:
I think so, however I wanted to avoid ptb constructions for better clarity

Reply:
Done